src/storage/concurrency_manager/key_handle.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::lock_table::LockTable;
+
+use parking_lot::Mutex;
+use std::{mem, sync::Arc};
+use tokio::sync::{Mutex as AsyncMutex, MutexGuard as AsyncMutexGuard};
+use txn_types::{Key, Lock};
+
+/// An entry in the in-memory table providing functions related to a specific
+/// key.
+pub struct KeyHandle {
+    pub key: Key,
+    table: LockTable,
+    mutex: AsyncMutex<()>,
+    lock_store: Mutex<Option<Lock>>,
+}
+
+impl KeyHandle {
+    pub fn new(key: Key, table: LockTable) -> Self {
+        KeyHandle {
+            key,
+            table,
+            mutex: AsyncMutex::new(()),
+            lock_store: Mutex::new(None),
+        }
+    }
+
+    pub async fn lock(self: Arc<Self>) -> KeyHandleGuard {
+        // Safety: `_mutex_guard` is declared before `handle_ref` in `KeyHandleGuard`.
+        // So the mutex guard will be released earlier than the `Arc<KeyHandle>`.
+        // Then we can make sure the mutex guard doesn't point to released memory.
+        let mutex_guard = unsafe { mem::transmute(self.mutex.lock().await) };
+        KeyHandleGuard {
+            _mutex_guard: mutex_guard,
+            handle: self,
+        }
+    }
+
+    pub fn with_lock<T>(&self, f: impl FnOnce(&Option<Lock>) -> T) -> T {
+        f(&*self.lock_store.lock())
+    }
+}
+
+impl Drop for KeyHandle {
+    fn drop(&mut self) {
+        self.table.remove(&self.key);
+    }
+}
+
+/// A `KeyHandle` with its mutex locked.
+pub struct KeyHandleGuard {
+    // It must be declared before `handle_ref` so it will be dropped before
+    // `handle_ref`.
+    _mutex_guard: AsyncMutexGuard<'static, ()>,
+    handle: Arc<KeyHandle>,
```

Comment:
Should add a note that it is unsafe to mutate handle to point at a different KeyHandle